admin/admin.go

```diff
@@ -5,231 +5,116 @@
-	AddBucket(namespace string, b *pb.BucketConfig) error
-	UpdateBucket(namespace string, b *pb.BucketConfig) error
+const (
+	LogPattern = "%s - - [%s] \"%s\" %d %d %f\n"
+)
 
-	DeleteNamespace(namespace string) error
-	AddNamespace(n *pb.NamespaceConfig) error
-	UpdateNamespace(n *pb.NamespaceConfig) error
+type ResponseWrapper struct {
+	http.ResponseWriter
+
+	ip            string
+	time          time.Time
+	method        string
+	uri           string
+	protocol      string
+	status        int
+	responseBytes int64
+	elapsedTime   time.Duration
 }
 
 // ServeAdminConsole serves up an admin console for an Administrable using Go's built-in HTTP server
 // library. `assetsDirectory` contains HTML templates and other UI assets. If empty, no UI will be
 // served, and only REST endpoints under `/api/` will be served.
-func ServeAdminConsole(a Administrable, mux *http.ServeMux, assetsDirectory string) {
-	logging.Print("Serving admin console.")
+func ServeAdminConsole(a Administrable, mux *http.ServeMux, assetsDirectory string, development bool) {
 	if assetsDirectory != "" {
-		files, err := ioutil.ReadDir(assetsDirectory)
-		check(err)
-		htmlFiles := make([]string, 0)
-		for _, f := range files {
-			if !f.IsDir() && strings.HasSuffix(f.Name(), ".html") {
-				htmlFiles = append(htmlFiles, assetsDirectory+"/"+f.Name())
-			}
-		}
-		mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
-			http.Redirect(w, r, "/admin/", 301)
-		})
-		mux.Handle("/admin/", &uiHandler{a, reloadTemplates(htmlFiles), htmlFiles})
-		mux.Handle("/js/", http.FileServer(http.Dir(assetsDirectory)))
+		logging.Printf("Serving assets from %s", assetsDirectory)
+
+		mux.Handle("/", loggingHandler(http.RedirectHandler("/admin/", 301)))
+		mux.Handle("/admin/", loggingHandler(NewUIHandler(a, assetsDirectory, development)))
+		mux.Handle("/js/", loggingHandler(http.FileServer(http.Dir(assetsDirectory))))
 	} else {
 		logging.Print("Not serving admin web UI.")
+		mux.Handle("/", loggingHandler(http.NotFoundHandler()))
 	}
-	mux.Handle("/api/", &apiHandler{a})
-}
 
-func reloadTemplates(files []string) *template.Template {
-	return template.Must(template.New("admin").ParseFiles(files...))
-}
+	apiNotFoundHandler := loggingHandler(apiRequestHandler(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		WriteJSONError(w, errors.New("NotFound"), http.StatusNotFound)
+	})))
 
-// uiHandler is an http.Handler for the web UI.
-type uiHandler struct {
-	a Administrable
-	t *template.Template
-	h []string
-}
+	mux.Handle("/api", apiNotFoundHandler)
+	mux.Handle("/api/", apiNotFoundHandler)
 
-func (h *uiHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-	// TODO(manik) remove this. Currently in place for testing. Shouldn't need to reload templates on every request.
-	h.t = reloadTemplates(h.h)
+	bucketsHandler := loggingHandler(apiRequestHandler(NewBucketsAPIHandler(a)))
+	mux.Handle("/api/buckets", bucketsHandler)
```

Comment:
I was thinking about this. I don't think `/api/buckets/...` and `/api/namespaces/...` is the way to go. I was going to remove that stuff and replace it with something like:

* `/api/` -- anything related to the global config. E.g., `GET /api/` would return the config for the entire system. PUT, DELETE and POST would affect the global bucket.
* `/api/${namespace}/` -- anything related to the specified namespace. `GET /api/${namespace}` would return the config for the specified namespace only. PUT, DELETE and POST would affect this namespace config.
* `/api/${namespace}/${bucket}` -- anything related to the specified bucket in the specified namespace.

That way, we only need a handler for `/api/`...